# Movie Reccomendation System

Data is taken from https://www.kaggle.com/tmdb/tmdb-movie-metadata <br>

Demographic Filtering <br>
Content Based Filtering <br>
Collaborative Filtering <br>

In [42]:
import numpy as np
import pandas as pd


In [43]:
df_credits = pd.read_csv("data/tmdb_5000_credits.csv")
df_movies = pd.read_csv("data/tmdb_5000_movies.csv")


## EDA

In [44]:
df_movies.shape,df_credits.shape

((4803, 20), (4803, 4))

In [45]:
df_movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count'],
      dtype='object')

In [46]:
df_credits.columns

Index(['movie_id', 'title', 'cast', 'crew'], dtype='object')

In [47]:
df_credits = df_credits.rename(columns={'movie_id':'id'})
df_credits.columns

Index(['id', 'title', 'cast', 'crew'], dtype='object')

#### Merging the movies and credits dataframe

In [48]:
df= pd.merge(df_credits,df_movies,on='id')
df.columns

Index(['id', 'title_x', 'cast', 'crew', 'budget', 'genres', 'homepage',
       'keywords', 'original_language', 'original_title', 'overview',
       'popularity', 'production_companies', 'production_countries',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'title_y', 'vote_average', 'vote_count'],
      dtype='object')

In [49]:
df.head()

id                                   title_x  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew     budget  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...  237000000   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  300000000   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  245000000   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  250000000   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  260000000   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage  \
0                   http://www.avatarmovie.com/   
1  http://disney.go.com/disneypictures/pirates/   
2   http://www.sonypictures.com/movies/spectre/   
3            http://www.thedarkknightrises.com/   
4          http://movies.disney.com/john-carter   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  ...  \
0                                    Avatar  ...   
1  Pirates of the Caribbean: At World's End  ...   
2                                   Spectre  ...   
3                     The Dark Knight Rises  ...   
4                               John Carter  ...   

                                production_countries  release_date  \
0  [{"iso_3166_1": "US", "name": "United States o...    2009-12-10   
1  [{"iso_3166_1": "US", "name": "United States o...    2007-05-19   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...    2015-10-26   
3  [{"iso_3166_1": "US", "name": "United States o...    2012-07-16   
4  [{"iso_3166_1": "US", "name": "United States o...    2012-03-07   

      revenue runtime                                   spoken_languages  \
0  2787965087   162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...   
1   961000000   169.0           [{"iso_639_1": "en", "name": "English"}]   
2   880674609   148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...   
3  1084939099   165.0           [{"iso_639_1": "en", "name": "English"}]   
4   284139100   132.0           [{"iso_639_1": "en", "name": "English"}]   

     status                                         tagline  \
0  Released                     Enter the World of Pandora.   
1  Released  At the end of the world, the adventure begins.   
2  Released                           A Plan No One Escapes   
3  Released                                 The Legend Ends   
4  Released            Lost in our world, found in another.   

                                    title_y vote_average vote_count  
0                                    Avatar          7.2      11800  
1  Pirates of the Caribbean: At World's End          6.9       4500  
2                                   Spectre          6.3       4466

In [50]:
len(df)

4803

In [51]:
df.columns

Index(['id', 'title_x', 'cast', 'crew', 'budget', 'genres', 'homepage',
       'keywords', 'original_language', 'original_title', 'overview',
       'popularity', 'production_companies', 'production_countries',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'status',
       'tagline', 'title_y', 'vote_average', 'vote_count'],
      dtype='object')

In [52]:
col_dict = df.dtypes.to_dict()

In [53]:
col_dict

{'id': dtype('int64'),
 'title_x': dtype('O'),
 'cast': dtype('O'),
 'crew': dtype('O'),
 'budget': dtype('int64'),
 'genres': dtype('O'),
 'homepage': dtype('O'),
 'keywords': dtype('O'),
 'original_language': dtype('O'),
 'original_title': dtype('O'),
 'overview': dtype('O'),
 'popularity': dtype('float64'),
 'production_companies': dtype('O'),
 'production_countries': dtype('O'),
 'release_date': dtype('O'),
 'revenue': dtype('int64'),
 'runtime': dtype('float64'),
 'spoken_languages': dtype('O'),
 'status': dtype('O'),
 'tagline': dtype('O'),
 'title_y': dtype('O'),
 'vote_average': dtype('float64'),
 'vote_count': dtype('int64')}

## Demographic Filtering

Using IMBD weighted rating <br>   wr= (v/v+m)*R + (m/v+m)*C <br>
where <br>  v = number of votes<br>
        m = minimum number of votes required to included in the list<br>
        R = average rating<br>
        C = mean votes across whole report<br>
        

In [54]:
df[['original_title','vote_average','vote_count']][:5]

original_title  vote_average  vote_count
0                                    Avatar           7.2       11800
1  Pirates of the Caribbean: At World's End           6.9        4500
2                                   Spectre           6.3        4466
3                     The Dark Knight Rises           7.6        9106
4                               John Carter           6.1        2124

In [55]:
C = df['vote_average'].mean()
C

6.092171559442011

In [56]:
#m = minimum number of votes required to included in the list<br>
m= df['vote_count'].quantile(0.9)
m

1838.4000000000015

In [57]:
#movies with selected
df1=df.copy().loc[df['vote_count'] >= m]
df1.shape

(481, 23)

In [58]:
#def weighted rating
def weighted_rating(x,m=m,C=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m)*R + m/(m+v)*C)

In [59]:
df1['score'] = df1.apply(weighted_rating,axis=1)
df1.shape

(481, 24)

In [60]:
df1['score']

0       7.050669
1       6.665696
2       6.239396
3       7.346721
4       6.096368
          ...   
4291    6.693677
4300    7.366378
4302    7.210428
4337    7.198026
4602    7.210563
Name: score, Length: 481, dtype: float64

In [61]:
df2 = df1.sort_values('score', ascending=False)
df2.head()

id                   title_x  \
1881    278  The Shawshank Redemption   
662     550                Fight Club   
65      155           The Dark Knight   
3232    680              Pulp Fiction   
96    27205                 Inception   

                                                   cast  \
1881  [{"cast_id": 3, "character": "Andy Dufresne", ...   
662   [{"cast_id": 4, "character": "The Narrator", "...   
65    [{"cast_id": 35, "character": "Bruce Wayne", "...   
3232  [{"cast_id": 2, "character": "Vincent Vega", "...   
96    [{"cast_id": 1, "character": "Dom Cobb", "cred...   

                                                   crew     budget  \
1881  [{"credit_id": "52fe4231c3a36847f800b153", "de...   25000000   
662   [{"credit_id": "55731b8192514111610027d7", "de...   63000000   
65    [{"credit_id": "55a0eb4a925141296b0010f8", "de...  185000000   
3232  [{"credit_id": "52fe4269c3a36847f801caa3", "de...    8000000   
96    [{"credit_id": "56e8462cc3a368408400354c", "de...  160000000   

                                                 genres  \
1881  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
662                       [{"id": 18, "name": "Drama"}]   
65    [{"id": 18, "name": "Drama"}, {"id": 28, "name...   
3232  [{"id": 53, "name": "Thriller"}, {"id": 80, "n...   
96    [{"id": 28, "name": "Action"}, {"id": 53, "nam...   

                                          homepage  \
1881                                           NaN   
662     http://www.foxmovies.com/movies/fight-club   
65    http://thedarkknight.warnerbros.com/dvdsite/   
3232                                           NaN   
96           http://inceptionmovie.warnerbros.com/   

                                               keywords original_language  \
1881  [{"id": 378, "name": "prison"}, {"id": 417, "n...                en   
662   [{"id": 825, "name": "support group"}, {"id": ...                en   
65    [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
3232  [{"id": 396, "name": "transporter"}, {"id": 14...                en   
96    [{"id": 1014, "name": "loss of lover"}, {"id":...                en   

                original_title  ... release_date     revenue runtime  \
1881  The Shawshank Redemption  ...   1994-09-23    28341469   142.0   
662                 Fight Club  ...   1999-10-15   100853753   139.0   
65             The Dark Knight  ...   2008-07-16  1004558444   152.0   
3232              Pulp Fiction  ...   1994-10-08   213928762   154.0   
96                   Inception  ...   2010-07-14   825532764   148.0   

                                       spoken_languages    status  \
1881           [{"iso_639_1": "en", "name": "English"}]  Released   
662            [{"iso_639_1": "en", "name": "English"}]  Released   
65    [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
3232  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
96    [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                                                tagline  \
1881  Fear can hold you prisoner. Hope can set you f...   
662                             Mischief. Mayhem. Soap.   
65                                      Why So Serious?   
3232  Just because you are a character doesn't mean ...   
96                 Your mind is the scene of the crime.   

                       title_y vote_average vote_count     score  
1881  The Shawshank Redemption          8.5       8205  8.059258  
662                 Fight Club          8.3       9413  7.939256  
65             The Dark Knight          8.2      12002  7.920020  
3232              Pulp Fiction          8.3       8428  7.904645  
96                   Inception          8.1      13752  7.863239  

[5 rows x 24 columns]

In [62]:
df2[['original_title','vote_average','vote_count','score']][0:10]


original_title  vote_average  vote_count  \
1881                       The Shawshank Redemption           8.5        8205   
662                                      Fight Club           8.3        9413   
65                                  The Dark Knight           8.2       12002   
3232                                   Pulp Fiction           8.3        8428   
96                                        Inception           8.1       13752   
3337                                  The Godfather           8.4        5893   
95                                     Interstellar           8.1       10867   
809                                    Forrest Gump           8.2        7927   
329   The Lord of the Rings: The Return of the King           8.1        8064   
1990                        The Empire Strikes Back           8.2        5879   

         score  
1881  8.059258  
662   7.939256  
65    7.920020  
3232  7.904645  
96    7.863239  
3337  7.851236  
95    7.809479  
809   7.803188  
329   7.727243  
1990  7.697884

## Content Based Filtering

### Plot Based Reccomendation 

In [63]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(stop_words='english')

In [64]:
df['overview'][0:5]

0    In the 22nd century, a paraplegic Marine is di...
1    Captain Barbossa, long believed to be dead, ha...
2    A cryptic message from Bond’s past sends him o...
3    Following the death of District Attorney Harve...
4    John Carter is a war-weary, former military ca...
Name: overview, dtype: object

In [65]:
df['overview'].isna().sum()

3

In [66]:
df['overview'] = df['overview'].fillna('')

In [67]:
df['overview'].isna().sum()

0

In [68]:
tfidf_matrix = tfidf.fit_transform(df['overview'])
tfidf_matrix.shape

(4803, 20978)

In [69]:
df.shape

(4803, 23)

#### We created a TF-IDF matrix for the 4803 movies rows and 20978 words as columns

### Cosine Similarity

In [70]:
from sklearn.metrics.pairwise import linear_kernel
cosine_sim = linear_kernel(tfidf_matrix,tfidf_matrix)

In [71]:
#getting indices of movies 
indices = pd.Series(df.index, index = df['title_x']).drop_duplicates()

In [72]:
def get_reccomendation(title,cosine_sim=cosine_sim):
    idx = indices[title]
    sim_score = list(enumerate(cosine_sim[idx]))
    sim_score = sorted(sim_score, key = lambda x : x[1], reverse = True)
    sim_score = sim_score[1:11]
    movie_indices = [i[0] for i in sim_score]
    return df['title_x'].iloc[movie_indices]

In [73]:
get_reccomendation('The Dark Knight Rises')

65                              The Dark Knight
299                              Batman Forever
428                              Batman Returns
1359                                     Batman
3854    Batman: The Dark Knight Returns, Part 2
119                               Batman Begins
2507                                  Slow Burn
9            Batman v Superman: Dawn of Justice
1181                                        JFK
210                              Batman & Robin
Name: title_x, dtype: object

In [74]:
get_reccomendation('The Godfather')

2731     The Godfather: Part II
1873                 Blood Ties
867     The Godfather: Part III
3727                 Easy Money
3623                       Made
3125                     Eulogy
3896                   Sinister
4506            The Maid's Room
3783                        Joe
2244      The Cold Light of Day
Name: title_x, dtype: object

### Credits , Keywords and Genre based Reccomendation

In [75]:
df.crew.dtype

dtype('O')

In [76]:
from ast import literal_eval

features = ['cast','crew','keywords','genres']
for feature in features:
    df[feature] = df[feature].apply(literal_eval)

In [105]:
df.cast.dtype

dtype('O')

In [77]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [78]:
# Returns the list top 3 elements or entire list; whichever is more.
def get_list(x):
    if isinstance(x, list):
        names = [i['name'] for i in x]
        #Check if more than 3 elements exist. If yes, return only first three. If no, return entire list.
        if len(names) > 3:
            names = names[:3]
        return names

    #Return empty list in case of missing/malformed data
    return []

In [80]:
df['director'] = df['crew'].apply(get_director)

features = ['cast','keywords','genres']

for feature in features:
    df[feature] = df[feature].apply(get_list)

In [83]:
# Print the new features of the first 3 films
df[['title_x', 'cast', 'director', 'keywords', 'genres']].head(3)

title_x  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                               cast        director  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   James Cameron   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  Gore Verbinski   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]      Sam Mendes   

                              keywords                        genres  
0   [culture clash, future, space war]  [Action, Adventure, Fantasy]  
1   [ocean, drug abuse, exotic island]  [Adventure, Fantasy, Action]  
2  [spy, based on novel, secret agent]    [Action, Adventure, Crime]

In [84]:
def clean_data(x):
    if isinstance(x,list):
        return [str.lower(i.replace(" ","")) for i in x]
    else:
        if isinstance(x,str):
            return str.lower(x.replace(" ",""))
        else:
            return ''

In [89]:
# cleaning data using clean_data

features = ['cast','keywords','director','genres']

for feature in features:
    df[feature] = df[feature].apply(clean_data)

In [94]:
def create_soup(x):
    return ' '.join(x['keywords']) + ' ' + ' '.join(x['cast']) + ' ' + x['director'] + ' ' + ' '.join(x['genres'])
df['soup'] = df.apply(create_soup, axis=1)

In [95]:
# Import CountVectorizer and create the count matrix
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df['soup'])

In [96]:
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim2 = cosine_similarity(count_matrix, count_matrix)

In [99]:
#Reset index of our main DataFrame and construct reverse mapping as before
df = df.reset_index()
indices = pd.Series(df.index, index=df['title_x'])

In [103]:
get_reccomendation('The Dark Knight Rises', cosine_sim2)


65               The Dark Knight
119                Batman Begins
4638    Amidst the Devil's Wings
1196                The Prestige
3073           Romeo Is Bleeding
3326              Black November
1503                      Takers
1986                      Faster
303                     Catwoman
747               Gangster Squad
Name: title_x, dtype: object

In [104]:
get_reccomendation('The Godfather', cosine_sim2)


867      The Godfather: Part III
2731      The Godfather: Part II
4638    Amidst the Devil's Wings
2649           The Son of No One
1525              Apocalypse Now
1018             The Cotton Club
1170     The Talented Mr. Ripley
1209               The Rainmaker
1394               Donnie Brasco
1850                    Scarface
Name: title_x, dtype: object

## Collabartiv Filtering